### PR TITLE
[BERT] dataset preparation script for pre-training

### PR DESCRIPTION
**Pull request by haven-jeon**
## Description ##
- Simple script convertion from https://github.com/google-research/bert/blob/master/create_pretraining_data.py to Gluon(MXRecordIO).

### Essentials ###
- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by haven-jeon** [comment]
#474

**Comment by mli** [comment]
Job PR-489/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-489/2/index.html

File: scripts/bert/create_pretraining_data.py [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+    default=0.1,
+    help='Probability of creating sequences which are shorter than the '
+    'maximum length.')
+
+args = parser.parse_args()
+
+logging.info(args)
+
+
+class TrainingInstance:
```
Comment (eric-haibin-lin) [inline scripts/bert/create_pretraining_data.py]: inherit from object for py2 compatibility

File: scripts/bert/create_pretraining_data.py [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+
+    writer_index = 0
+
+    total_written = 0
+    for (inst_index, instance) in enumerate(instances):
+        input_ids = tokenizer.convert_tokens_to_ids(instance.tokens)
+        input_mask = [1] * len(input_ids)
+        segment_ids = list(instance.segment_ids)
+        assert len(input_ids) <= max_seq_length
+
```
Comment (eric-haibin-lin) [inline scripts/bert/create_pretraining_data.py]: For padded data, for consistency let's put 
```
pad = tokenizer.convert_tokens_to_ids('[PAD]')
...
            input_ids.append(pad)
            input_mask.append(pad)
            segment_ids.append(pad)

```

File: scripts/bert/create_pretraining_data.py [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+            del trunc_tokens[0]
+        else:
+            trunc_tokens.pop()
+
+
+def main():
+    """
+    main function
+    """
+    vocab_obj = pickle.load(open(args.vocab_file, 'rb'))
```
Comment (eric-haibin-lin) [inline scripts/bert/create_pretraining_data.py]: Would it be better to assume users have some vocab serialized with `file.write(vocab.to_json())`? If so we should use `Vocab.from_json(open(args.vocab_file).read())`? 
This way user can also provide the path to auto-downloaded vocab from pretrained BERT

**Comment by eric-haibin-lin** [comment]
Could you also provide some example code how these recordIO can be read for training?

File: scripts/bert/create_pretraining_data.py [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+        features['input_mask'] = input_mask
+        features['segment_ids'] = segment_ids
+        features['masked_lm_positions'] = masked_lm_positions
+        features['masked_lm_ids'] = masked_lm_ids
+        features['masked_lm_weights'] = masked_lm_weights
+        features['next_sentence_labels'] = [next_sentence_label]
+
+        example = features
+
+        writers[writer_index].write(pickle.dumps(example))
```
Comment (leezu) [inline scripts/bert/create_pretraining_data.py]: It might be inefficient to store a RecordIO file of pickled entries? Pickling is very expensive, and is now applied to every sentence individuall during save & load.

If the data fits in memory, consider using a numpy array for serialization. It should allow speeding things up by more than an order of magnitude. When dealing with variable length data, you must concatenate the variable length arrays to a single array and remember the concatenation-points so that you can break up the array again when loading it. See eg https://github.com/dmlc/gluon-nlp/blob/8ec91d8/scripts/machine_translation/dataprocessor.py#L33-L65

What do you think?

**Comment by mli** [comment]
Job PR-489/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-489/3/index.html

File: scripts/bert/create_pretraining_data.py (hunk 4 above)
Comment (leezu) [reply scripts/bert/create_pretraining_data.py]: (Regarding the code I linked above, it would be simpler to use `numpy.split` instead of the list comprehension and use of `zip` in the load function.)

Comment (haven-jeon) [reply scripts/bert/create_pretraining_data.py]: Thanks for comments . Also, there is several performance issues to store the duplicated keys for every records.

File: scripts/bert/create_pretraining_data.py (hunk 1 above)
Comment (haven-jeon) [reply scripts/bert/create_pretraining_data.py]: Sure!

File: scripts/bert/create_pretraining_data.py (hunk 3 above)
Comment (haven-jeon) [reply scripts/bert/create_pretraining_data.py]: It needs to follow json  for object serialization. Thanks.

File: scripts/bert/create_pretraining_data.py (hunk 2 above)
Comment (haven-jeon) [reply scripts/bert/create_pretraining_data.py]: Can we be sure padding of  input_mask, segment_ids will be same?

input_mask, segment_ids will have only '0', '1'

This will be correct.

```
pad = tokenizer.convert_tokens_to_ids('[PAD]')
...
            input_ids.append(pad)
            input_mask.append(0)
            segment_ids.append(0)
```

File: scripts/bert/create_pretraining_data.py (hunk 4 above)
Comment (leezu) [reply scripts/bert/create_pretraining_data.py]: Possibly https://github.com/dmlc/gluon-nlp/pull/498 can be extended (or similar) to take care of and hide the np.concatenate & np.split

**Comment by mli** [comment]
Job PR-489/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-489/4/index.html

File: scripts/bert/create_pretraining_data.py (hunk 2 above)
Comment (eric-haibin-lin) [reply scripts/bert/create_pretraining_data.py]: Oh you're right. mask and seg_id should remain 0

**Comment by mli** [comment]
Job PR-489/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-489/6/index.html

**Comment by eric-haibin-lin** [comment]
I'm in favor of merging this PR first. I made some incremental enhancement to support numpy output in #504
